feat(tabs): add animation when switching tabs, include optional dynamic height

andrewseguin opened the pull request.

hansl commented on src/lib/tabs/_tabs-common.scss:
```diff
@@ -36,5 +36,5 @@ $md-tab-bar-height: 48px !default;
   height: 2px;
-  transition: 350ms ease-out;
+  transition: 500ms ease-out;
 }
```
End of file, end of line.

hansl commented on src/lib/tabs/tab-body.html:
```diff
@@ -0,0 +1,6 @@
+     (@position.start)="_onAnimationStarted($event)"
+     (@position.done)="_onAnimationComplete($event)">
+  <template portalHost></template>
+</div>
```
End of line.

kara commented on src/lib/tabs/tab-group.spec.ts:
```diff
@@ -136,6 +139,53 @@ describe('MdTabGroup', () => {
+
+    it('should update tab positions and attach content when selected', fakeAsync(() => {
+      fixture.detectChanges();
+      let tabComponent = fixture.debugElement.query(By.css('md-tab-group')).componentInstance;
```
Out of curiosity, why is this one `let` while the other is `const`? Seems like they could both be `const`.

andrewseguin replied:
Thanks, I tend to just default to `let` rather than `const`. Gotta break that habit.

kara commented on src/lib/tabs/tab-body.html:
```diff
@@ -0,0 +1,6 @@
+<div class="md-tab-body-content"
+     [@position]="_position"
```
Nit: This animation name doesn't really describe what is happening with the animation.

kara commented on src/lib/tabs/tabs.ts:
```diff
@@ -30,6 +37,7 @@ import {MdTabNavBar, MdTabLink} from './tab-nav-bar/tab-nav-bar';
 import {MdInkBar} from './ink-bar';
 import {Observable} from 'rxjs/Observable';
 import 'rxjs/add/operator/map';
+import {PortalHostDirective} from '../core/portal/portal-directives';
```
It might make more sense to add this to the existing import list on line 19 from `../core`. I think it's exported there too.

hansl commented on src/lib/tabs/tabs.ts:
```diff
@@ -248,15 +249,123 @@ export class MdTabGroup {
+  templateUrl: 'tab-body.html',
+  animations: [
+    trigger('position', [
+      state('left', style({transform: 'translateX(-100%)'})),
```
According to https://jsperf.com/translate3d-vs-xy/28, `translate3d` is the fastest in most browsers (because 3d rendering), which is why we use it in most places.

andrewseguin replied:
Seems that Chrome has them set at about equal operations but 3D is way faster in Safari (and probably others). Thanks for the heads up

kara commented on src/lib/tabs/tabs.ts:
```diff
@@ -248,15 +249,123 @@ export class MdTabGroup {
+   * Sets the height of the body wrapper to the height of the activating tab if dynamic
+   * height property is true.
+   */
+  _setTabBodyWrapperHeight(e: number) {
```
Nit: return type?

kara replied:
Maybe `height` would be more descriptive than `e`? It's not clear at first glance what it stands for.

andrewseguin commented on src/lib/tabs/tab-group.scss:
```diff
@@ -32,19 +32,28 @@ md-ink-bar {
-    display: block;
+md-tab-body {
+  display: block;
+  position: absolute;
```
I don't see that we have one already. Want me to make one in this change?

kara replied:
We do have md-fullscreen, which is I think what Hans means. But it is position: fixed, not absolute.

hansl replied:
Right, there should probably be `md-fill` mixin or something like that with `position: absolute`. That mixin does not work right now.

andrewseguin commented on src/lib/tabs/tabs.ts:
```diff
@@ -248,15 +249,123 @@ export class MdTabGroup {
+    ])
+  ]
+})
+export class MdTabBody implements OnInit {
```
Primarily code organization - portal attach/detach + animations would have added quite a bit to tab group and I don't think it would have been as readable about what is happening between all the interactions

kara replied:
Okay, that's fair.  Thought that might be the case.

andrewseguin commented:
Changes made in response to the comments. Thanks for reviewing

andrewseguin commented:
Switched md-fullscreen back to using absolute positioning (changed in #893) which is safe for its use in sidenav. Renamed as md-fill and used in the tab body.

andrewseguin commented:
Sorry, added one more small change to support RTL animation direction

hansl commented:
LGTM.

jelbourn commented on e2e/components/tabs/tabs.e2e.ts:
```diff
@@ -77,8 +77,17 @@ function getFocusStates(elements: ElementArrayFinder) {
+/**
+ * Returns an array of true/false that represents the active states for the provided elements
+ * @param elements
+ * @returns {webdriver.promise.Promise<Promise<boolean>[]>|webdriver.promise.Promise<T[]>}
```
Since this is TypeScript, you don't need the type in the JsDoc (these were formerly js)

jelbourn commented on src/lib/tabs/tabs.ts:
```diff
@@ -248,15 +253,132 @@ export class MdTabGroup {
+    } else if (v > 0) {
+      this._position = this.getLayoutDirection() == 'ltr' ? 'right' : 'left';
+    }
+    if (v == 0) { this._position = 'center'; }
```
Just `else` here?

jelbourn commented on src/lib/tabs/tab-group.spec.ts:
```diff
@@ -136,6 +139,53 @@ describe('MdTabGroup', () => {
       expect(component.selectEvent.index).toBe(2);
     }));
+
+    it('should update tab positions and attach content when selected', fakeAsync(() => {
```
Also a test for RTL?

andrewseguin commented:
Will make a note to check out screen reader support after this PR
